README.md
# Very short description of the package

[![Latest Version on Packagist](https://img.shields.io/packagist/v/strebl/inetworx-client.svg?style=flat-square)](https://packagist.org/packages/strebl/inetworx-client)
[![Build Status](https://img.shields.io/travis/strebl/inetworx-client/master.svg?style=flat-square)](https://travis-ci.org/strebl/inetworx-client)
[![StyleCI](https://styleci.io/repos/91091045/shield)](https://styleci.io/repos/91091045)
[![SensioLabsInsight](https://img.shields.io/sensiolabs/i/4662f707-185d-476e-a110-0c34feb41169.svg?style=flat-square)](https://insight.sensiolabs.com/projects/4662f707-185d-476e-a110-0c34feb41169)
[![Quality Score](https://img.shields.io/scrutinizer/g/strebl/inetworx-client.svg?style=flat-square)](https://scrutinizer-ci.com/g/strebl/inetworx-client)
[![Total Downloads](https://img.shields.io/packagist/dt/strebl/inetworx-client.svg?style=flat-square)](https://packagist.org/packages/strebl/inetworx-client)

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what PSRs you support to avoid any confusion with users and contributors.

## Installation

You can install the package via composer:

```bash
composer require strebl/inetworx-client
```


## Laravel

If you are using Laravel, you can register the service provider:

```php
'providers' => [
    // ...
    Strebl\Inetworx\InetworxServiceProvider::class,
];
```

To publish the config file to `config/inetworx.php` run (optional):

```bash
php artisan vendor:publish --provider="Strebl\Inetworx\InetworxServiceProvider"
```

If you are using Laravel, you can register the service provider:

```php
'aliases' => [
    // ...
    'Inetworx' => Strebl\Inetworx\InetworxFacade::class,
];
```

Set the environment variables with the correct values:

```
INETWORX_AUTH_HEADER_USERNAME=null
INETWORX_AUTH_HEADER_PASSWORD=null

INETWORX_API_USERNAME=null
INETWORX_API_PASSWORD=null
```



## Usage

### With Laravel

``` php
$skeleton = app(Strebl\Inetworx::class);
$sms->send($phoneNumber, 'Hello, Manuel!', $senderPhoneNumber);
```

Or you can use the Facade:

``` php
\Inetworx::send($phoneNumber, 'Hello, Manuel!', $senderPhoneNumber);
```

### Without Laravel

``` php
$sms = new Strebl\Inetworx(
    $authHeaderUsername,
    $authHeaderPassword,
    $apiUsername,
    $apiPassword,
);
$sms->send($phoneNumber, 'Hello, Manuel!', $senderPhoneNumber);
```

### Send a SMS

``` php
$sms->send($to, $message, $from);
```

### Get the remaining SMS credits

``` php
$sms->credit();
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email manuel@strebel.xyz instead of using the issue tracker.

## Credits

- [Manuel Strebel](https://github.com/strebl)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `inetworx-client` will be documented in this file

## 1.0.1 - 2017-05-15

- PHP 7.0 support

## 1.0.0 - 2017-05-15

- initial release


tests/Integration/TestCase.php
<?php

namespace Strebl\Inetworx\Test\Integration;

use Strebl\Inetworx\InetworxFacade;
use Strebl\Inetworx\InetworxServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

abstract class TestCase extends Orchestra
{
    public function setUp()
    {
        parent::setUp();
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            InetworxServiceProvider::class,
        ];
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageAliases($app)
    {
        return [
            'Inetworx' => InetworxFacade::class,
        ];
    }
}


tests/Integration/InetworxClientTest.php
<?php

namespace Strebl\Inetworx\Test\Integration;

use PHPUnit\Framework\TestCase;
use Strebl\Inetworx\InetworxClient;

class InetworxClientTest extends TestCase
{
    /**
     * @var InetworxClient
     */
    protected $client;

    public function setUp()
    {
        parent::setUp();

        $this->client = new InetworxClient(
            getenv('INETWORX_AUTH_HEADER_USERNAME'),
            getenv('INETWORX_AUTH_HEADER_PASSWORD'),
            getenv('INETWORX_API_USERNAME'),
            getenv('INETWORX_API_PASSWORD')
        );
    }

    /** @test */
    public function it_fetches_the_remaining_credit()
    {
        $credit = $this->client->credit();

        $this->assertInternalType('int', $credit);
        $this->assertGreaterThan(1, $credit);
    }

    /**
     * @test
     * @group costsMoney
     */
    public function it_sends_a_message()
    {
        $response = $this->client->send(getenv('TEST_NUMBER'), 'Test Message', 'Strebl');

        $this->assertEquals(200, $response['statusCode']);
    }
}


tests/Integration/InetworxServiceProviderTest.php
<?php

namespace Strebl\Inetworx\Test\Integration;

use Inetworx;

class InetworxServiceProviderTest extends TestCase
{
    /** @test */
    public function test_the_facade()
    {
        $credit = Inetworx::credit();

        $this->assertInternalType('int', $credit);
        $this->assertGreaterThan(1, $credit);
    }
}


src/InetworkxClient.php
<?php

namespace Strebl\Inetworx;

use GuzzleHttp\Client;
use Strebl\Inetworx\Exceptions\ApiErrorException;

class InetworkxClient
{
    /**
     * @var string
     */
    protected $baseUrl = 'https://sms.inetworx.ch/';

    /**
     * @var string
     */
    protected $broadcastUrl = 'smsapp/sendsms.php';

    /**
     * @var string
     */
    protected $optionsUrl = 'smsapp/options.php';

    /**
     * @var Client
     */
    protected $httpClient;

    /**
     * @var string
     */
    protected $apiPassword;

    /**
     * @var string
     */
    protected $apiUsername;

    /**
     * Inetworx Client Constructor.
     *
     * @param string $authHeaderUsername
     * @param string $authHeaderPassword
     * @param string $apiUsername
     * @param string $apiPassword
     * @param $client
     */
    public function __construct($authHeaderUsername, $authHeaderPassword, $apiUsername, $apiPassword, $client = null)
    {
        if ($client) {
            $this->httpClient = $client;
        } else {
            $this->createHttpClient($authHeaderUsername, $authHeaderPassword);
        }

        $this->apiUsername = $apiUsername;
        $this->apiPassword = $apiPassword;
    }

    /**
     * Send the message.
     *
     * @param $phoneNumber
     * @param $message
     * @param $fromNumber
     *
     * @return array $response
     */
    public function send($phoneNumber, $message, $fromNumber)
    {
        return $this->sendRequest($this->getBroadcastUrl(), [
            'sender'  => $fromNumber,
            'rcpt'    => $phoneNumber,
            'msgbody' => $message,
        ]);
    }

     /**
      * Get the remaining SMS credits.
      *
      * @throws ApiErrorException
      *
      * @return int $credit
      */
     public function credit()
     {
         return (int) $this->sendRequest($this->getOptionsUrl(), ['option'  => 'quota'])['body'];
     }

    /**
     * @param $url
     * @param $params
     *
     * @throws ApiErrorException
     *
     * @return array
     */
    protected function sendRequest($url, $params)
    {
        $response = $this
             ->getHttpClient()
             ->post($url, [
                 'form_params' => array_merge([
                     'user'    => $this->apiUsername,
                     'pass'    => $this->apiPassword,
                 ], $params),
             ]);

        if ($response->getStatusCode() !== 200 || $this->responseCodeFromBody($response) !== 200) {
            throw new ApiErrorException(
                 $this->responseBody($response)
             );
        }

        return [
             'body'          => $this->responseBody($response),
             'full_body'     => (string) $response->getBody(),
             'statusCode'    => $response->getStatusCode(),
             'statusMessage' => $response->getReasonPhrase(),
         ];
    }

    /**
     * The Inetworx API doesn't respond with a useful HTTP Status Code if the request fails.
     * It puts the status code into the response body.
     * for example: 404: Invalid Option.
     *
     * @return int $code
     */
    protected function responseCodeFromBody($response)
    {
        return (int) explode(':', (string) $response->getBody())[0];
    }

    /**
     * Inetworx puts the status code in the response body too.
     * Because of that, we have to parse the answer.
     *
     * @param $response
     *
     * @return string $body
     */
    protected function responseBody($response)
    {
        return trim(explode(':', (string) $response->getBody())[1]);
    }

    /**
     * @return string
     */
    public function getBaseUrl()
    {
        return $this->baseUrl;
    }

    /**
     * @param string $baseUrl
     */
    public function setBaseUrl($baseUrl)
    {
        $this->baseUrl = $baseUrl;
    }

    /**
     * @return string
     */
    public function getBroadcastUrl()
    {
        return $this->broadcastUrl;
    }

    /**
     * @param string $broadcastUrl
     */
    public function setBroadcastUrl($broadcastUrl)
    {
        $this->broadcastUrl = $broadcastUrl;
    }

    /**
     * @return string
     */
    public function getOptionsUrl()
    {
        return $this->optionsUrl;
    }

    /**
     * @param string $optionsUrl
     */
    public function setOptionsUrl($optionsUrl)
    {
        $this->optionsUrl = $optionsUrl;
    }

    /**
     * @return Client
     */
    protected function getHttpClient()
    {
        return $this->httpClient;
    }

    /**
     * @param $client
     *
     * @return void
     */
    protected function setHttpClient($client)
    {
        $this->httpClient = $client;
    }

    /**
     * @return void
     */
    protected function createHttpClient($authHeaderUsername, $authHeaderPassword)
    {
        $this->httpClient = new Client([
            'base_uri' => $this->getBaseUrl(),
            'auth'     => [
                $authHeaderUsername,
                $authHeaderPassword,
            ],
        ]);
    }
}


src/InetworxServiceProvider.php
<?php

namespace Strebl\Inetworx;

use Illuminate\Support\ServiceProvider;

class InetworxServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/inetworx.php' => config_path('inetworx.php'),
            ], 'config');
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        $this->mergeConfigFrom(__DIR__.'/../config/inetworx.php', 'inetworx');

        $this->app->bind(InetworxClient::class, function () {
            return new InetworxClient(
                config('inetworx.auth.credentials.auth_header.username'),
                config('inetworx.auth.credentials.auth_header.password'),
                config('inetworx.auth.credentials.api.username'),
                config('inetworx.auth.credentials.api.password')
            );
        });

        $this->app->alias(InetworxClient::class, 'inetworx-client');
    }
}


src/InetworxClient.php
<?php

namespace Strebl\Inetworx;

use GuzzleHttp\Client;
use Strebl\Inetworx\Exceptions\ApiErrorException;

class InetworxClient
{
    /**
     * @var string
     */
    protected $baseUrl = 'https://sms.inetworx.ch/';

    /**
     * @var string
     */
    protected $broadcastUrl = 'smsapp/sendsms.php';

    /**
     * @var string
     */
    protected $optionsUrl = 'smsapp/options.php';

    /**
     * @var Client
     */
    protected $httpClient;

    /**
     * @var string
     */
    protected $apiPassword;

    /**
     * @var string
     */
    protected $apiUsername;

    /**
     * Inetworx Client Constructor.
     *
     * @param string $authHeaderUsername
     * @param string $authHeaderPassword
     * @param string $apiUsername
     * @param string $apiPassword
     * @param $client
     */
    public function __construct($authHeaderUsername, $authHeaderPassword, $apiUsername, $apiPassword, $client = null)
    {
        if ($client) {
            $this->httpClient = $client;
        } else {
            $this->createHttpClient($authHeaderUsername, $authHeaderPassword);
        }

        $this->apiUsername = $apiUsername;
        $this->apiPassword = $apiPassword;
    }

    /**
     * Send the message.
     *
     * @param $phoneNumber
     * @param $message
     * @param $fromNumber
     *
     * @return array $response
     */
    public function send($phoneNumber, $message, $fromNumber)
    {
        return $this->sendRequest($this->getBroadcastUrl(), [
            'sender'  => $fromNumber,
            'rcpt'    => $phoneNumber,
            'msgbody' => $message,
            'coding'  => 2,
        ]);
    }

     /**
      * Get the remaining SMS credits.
      *
      * @throws ApiErrorException
      *
      * @return int $credit
      */
     public function credit()
     {
         return (int) $this->sendRequest($this->getOptionsUrl(), ['option'  => 'quota'])['body'];
     }

    /**
     * @param $url
     * @param $params
     *
     * @throws ApiErrorException
     *
     * @return array
     */
    protected function sendRequest($url, $params)
    {
        $response = $this
             ->getHttpClient()
             ->post($url, [
                 'form_params' => array_merge([
                     'user'    => $this->apiUsername,
                     'pass'    => $this->apiPassword,
                 ], $params),
             ]);

        if ($response->getStatusCode() !== 200 || $this->responseCodeFromBody($response) !== 200) {
            throw new ApiErrorException(
                 $this->responseBody($response)
             );
        }

        return [
             'body'      => $this->responseBody($response),
             'full_body' => (string) $response->getBody(),
             'status'    => $response->getStatusCode(),
             'message'   => $response->getReasonPhrase(),
         ];
    }

    /**
     * The Inetworx API doesn't respond with a useful HTTP Status Code if the request fails.
     * It puts the status code into the response body.
     * for example: 404: Invalid Option.
     *
     * @return int $code
     */
    protected function responseCodeFromBody($response)
    {
        return (int) explode(':', (string) $response->getBody())[0];
    }

    /**
     * Inetworx puts the status code in the response body too.
     * Because of that, we have to parse the answer.
     *
     * @param $response
     *
     * @return string $body
     */
    protected function responseBody($response)
    {
        return trim(explode(':', (string) $response->getBody())[1]);
    }

    /**
     * @return string
     */
    public function getBaseUrl()
    {
        return $this->baseUrl;
    }

    /**
     * @param string $baseUrl
     */
    public function setBaseUrl($baseUrl)
    {
        $this->baseUrl = $baseUrl;
    }

    /**
     * @return string
     */
    public function getBroadcastUrl()
    {
        return $this->broadcastUrl;
    }

    /**
     * @param string $broadcastUrl
     */
    public function setBroadcastUrl($broadcastUrl)
    {
        $this->broadcastUrl = $broadcastUrl;
    }

    /**
     * @return string
     */
    public function getOptionsUrl()
    {
        return $this->optionsUrl;
    }

    /**
     * @param string $optionsUrl
     */
    public function setOptionsUrl($optionsUrl)
    {
        $this->optionsUrl = $optionsUrl;
    }

    /**
     * @return Client
     */
    protected function getHttpClient()
    {
        return $this->httpClient;
    }

    /**
     * @param $client
     *
     * @return void
     */
    protected function setHttpClient($client)
    {
        $this->httpClient = $client;
    }

    /**
     * @return void
     */
    protected function createHttpClient($authHeaderUsername, $authHeaderPassword)
    {
        $this->httpClient = new Client([
            'base_uri' => $this->getBaseUrl(),
            'auth'     => [
                $authHeaderUsername,
                $authHeaderPassword,
            ],
        ]);
    }
}
